test/images/agnhost/tcp-seq-rst/tcp.go

```diff
@@ -0,0 +1,189 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Partially copied from https://github.com/bowei/lighthouse/blob/master/pkg/probe/tcp.go
+
+package tcpseqrst
+
+import (
+	"bytes"
+	"encoding/binary"
+	"log"
+	"net"
+)
+
+const (
+	tcpHeaderSize = 20
+	tcpProtoNum   = 6
+)
+
+const (
+	// FIN is a TCP flag
+	FIN uint16 = 1 << iota
+	// SYN is a TCP flag
+	SYN
+	// RST is a TCP flag
+	RST
+	// PSH is a TCP flag
+	PSH
+	// ACK is a TCP flag
+	ACK
+	// URG is a TCP flag
+	URG
+	// ECE is a TCP flag
+	ECE
+	// CWR is a TCP flag
+	CWR
+	// NS is a TCP flag
+	NS
+)
+
+type tcpPacket struct {
+	SrcPort    uint16 // 0
+	DestPort   uint16 // 2
+	Seq        uint32 // 4
+	Ack        uint32 // 8
+	Flags      uint16 // 13
+	WindowSize uint16 // 14
+	Checksum   uint16 // 16
+	UrgentPtr  uint16 // 18
+	// 20
+}
+
+func (t *tcpPacket) decode(pkt []byte) ([]byte, error) {
+	err := binary.Read(bytes.NewReader(pkt), binary.BigEndian, t)
+	if err != nil {
+		return nil, err
+	}
+
+	return pkt[t.DataOffset():], nil
+}
+
+func (t *tcpPacket) DataOffset() int {
+	return int((t.Flags >> 12) * 4)
+}
+
+func (t *tcpPacket) FlagString() string {
+	out := ""
+
+	if t.Flags&FIN != 0 {
+		out += "FIN "
+	}
+	if t.Flags&SYN != 0 {
+		out += "SYN "
+	}
+	if t.Flags&RST != 0 {
+		out += "RST "
+	}
+	if t.Flags&PSH != 0 {
+		out += "PSH "
+	}
+	if t.Flags&ACK != 0 {
+		out += "ACK "
+	}
+	if t.Flags&URG != 0 {
+		out += "URG "
+	}
+	if t.Flags&ECE != 0 {
+		out += "ECE "
+	}
+	if t.Flags&CWR != 0 {
+		out += "CWR "
+	}
+	if t.Flags&NS != 0 {
+		out += "NS "
+	}
+
+	return out
+}
+
+func (t *tcpPacket) encode(src, dest net.IP, data []byte) []byte {
+	pkt := make([]byte, 20, 20+len(data))
+
+	encoder := binary.BigEndian
+	encoder.PutUint16(pkt, t.SrcPort)
+	encoder.PutUint16(pkt[2:], t.DestPort)
+	encoder.PutUint32(pkt[4:], t.Seq)
+	encoder.PutUint32(pkt[8:], t.Ack)
+	encoder.PutUint16(pkt[12:], t.Flags)
+	encoder.PutUint16(pkt[14:], t.WindowSize)
+	encoder.PutUint16(pkt[18:], t.UrgentPtr)
+
+	checksum := checksumTCP(src, dest, pkt[:tcpHeaderSize], data)
+	pkt[16] = uint8(checksum & 0xff)
+	pkt[17] = uint8(checksum >> 8)
+
+	pkt = append(pkt, data...)
+
+	return pkt
+}
+
+func checksumTCP(src, dest net.IP, tcpHeader, data []byte) uint16 {
+	log.Printf("calling checksumTCP: %v %v %v %v", src, dest, tcpHeader, data)
+	chk := &tcpChecksumer{}
+
+	// Encode pseudoheader.
+	chk.add(src.To4())
```

Comment:
I know this is incorrect for IPv6, but the test works, 
I think that the tcp checksum offloading fixes it :smiley:

Reply:
"The test works" how? If the packet gets dropped for having a bad TCP checksum then the e2e test will pass but you aren't actually testing that the iptables filter rule works.